lib/srv/uacc/uacc.h

```diff
@@ -0,0 +1,141 @@
+// This function does not perform any argument validation.
+static int uacc_add_utmp_entry(const char *utmp_path, const char *wtmp_path, const char *username, const char *hostname, const int32_t remote_addr_v6[4], const char *tty_name, const char *id, int32_t tv_sec, int32_t tv_usec) {
+    if (utmpname(SELECT_UTMP_PATH(utmp_path)) != 0) {
+        return UACC_UTMP_FAILED_TO_SELECT_FILE;
+    }
+    struct utmp entry;
+    entry.ut_type = USER_PROCESS;
+    strncpy((char*) &entry.ut_line, tty_name, UT_LINESIZE);
+    strncpy((char*) &entry.ut_id, id, sizeof(entry.ut_id));
+    entry.ut_pid = getpid();
+    strncpy((char*) &entry.ut_host, hostname, sizeof(entry.ut_host));
+    strncpy((char*) &entry.ut_user, username, sizeof(entry.ut_user));
+    entry.ut_session = 1;
+    entry.ut_tv.tv_sec = tv_sec;
+    entry.ut_tv.tv_usec = tv_usec;
+    memcpy(&entry.ut_addr_v6, &remote_addr_v6, sizeof(int32_t) * 4);
+    errno = 0;
+    setutent();
+    if (errno != 0) {
+        return UACC_UTMP_FAILED_OPEN;
+    }
+    if (pututline(&entry) == NULL) {
+        endutent();
+        return errno == EPERM || errno == EACCES ? UACC_UTMP_MISSING_PERMISSIONS : UACC_UTMP_WRITE_ERROR;
+    }
+    endutent();
+    updwtmp(SELECT_WTMP_PATH(wtmp_path), &entry);
+    return 0;
+}
+
+// Low level C function to mark a database entry as DEAD_PROCESS.
+// This function does not perform string argument validation.
+static int uacc_mark_utmp_entry_dead(const char *utmp_path, const char *wtmp_path, const char *tty_name) {
+    if (utmpname(SELECT_UTMP_PATH(utmp_path)) != 0) {
+        return UACC_UTMP_FAILED_TO_SELECT_FILE;
+    }
+    errno = 0;
+    setutent();
+    if (errno != 0) {
+        return UACC_UTMP_FAILED_OPEN;
+    }
+    struct utmp line;
+    strncpy((char*) &line.ut_line, tty_name, UT_LINESIZE);
+    struct utmp *entry_t = getutline(&line);
+    if (entry_t == NULL) {
+        return UACC_UTMP_READ_ERROR;
+    }
+    struct utmp entry;
+    memcpy(&entry, entry_t, sizeof(struct utmp));
+    entry.ut_type = DEAD_PROCESS;
+    errno = 0;
+    setutent();
+    if (errno != 0) {
+        return UACC_UTMP_FAILED_OPEN;
+    }
+    if (pututline(&entry) == NULL) {
+        endutent();
+        return errno == EPERM || errno == EACCES ? UACC_UTMP_MISSING_PERMISSIONS : UACC_UTMP_WRITE_ERROR;
+    }
+    endutent();
+    updwtmp(SELECT_WTMP_PATH(wtmp_path), &entry);
+    return 0;
+}
+
+// Low level C function to check the database for an entry for a given user.
+// This function does not perform string argument validation.
+static int uacc_has_entry_with_user(const char *utmp_path, const char *user) {
+    if (utmpname(SELECT_UTMP_PATH(utmp_path)) != 0) {
+        return UACC_UTMP_FAILED_TO_SELECT_FILE;
+    }
+    errno = 0;
+    setutent();
+    if (errno != 0) {
+        return UACC_UTMP_FAILED_OPEN;
+    }
+    struct utmp *entry = getutent();
+    while (entry != NULL) {
+        if (entry->ut_type == USER_PROCESS && strcmp(user, entry->ut_user) == 0) {
```

Comment:
Should this use `strncmp` since the utmp fields aren't necessarily null terminated? Worst case, it seems like this would read into the hostname field.